File: backend/ai/evidence.py
"""Evidence: what an AMP answer is made of, and how sure AMP is of it (ADR-0022).

Every figure the Copilot, the brief or an explorer shows is a `Fact`. A fact
carries its value AND the answer to "how do you know?", taken from one fixed
vocabulary, so the answer has the same meaning on every screen and in every
test:

    MEASURED FACT          read from recorded data: a count, a sum, a status
    DERIVED METRIC         a fixed formula over measured facts: OEE, a rate,
                           money from the tenant's own configured unit value
    CORRELATION            two measured things moved together; NOT a cause
    RULE-BASED ASSESSMENT  a deterministic threshold or hand-weighted score
                           (health bands, risk points). Never called ML.
    MODEL ESTIMATE         a trained model's output, with its validation status
    UNKNOWN                AMP does not know, and says so instead of guessing

A whole result has a data state, the honest-data vocabulary of the sprint:
NO DATA, PARTIAL DATA, NOT MEASURED, NOT CONFIGURED, INSUFFICIENT HISTORY and
MODEL NOT VALIDATED. None of them is ever turned into a zero. "No production
recorded" and "0% OEE" are different claims, and only one of them is true when
a gateway is down (ADR-0014).

The tool layer adds its refusals (NOT PERMITTED, NOT LICENSED, NOT FOUND,
INVALID ARGUMENTS, FAILED) so a refusal is a result the UI can render and the
evaluation can count, never an exception a model can talk its way around.

The root-cause labels live here too, so there is one list:
CAUSE CONFIRMED / LIKELY CONTRIBUTOR / CORRELATED EVENT / INSUFFICIENT EVIDENCE.

`frontend/lib/evidence.ts` mirrors these strings;
test_copilot_evidence_vocabulary.py fails the build if the two drift.
"""
import math
from dataclasses import dataclass, field

# ── Provenance: how a figure was obtained ───────────────────────────
MEASURED = "MEASURED FACT"
DERIVED = "DERIVED METRIC"
CORRELATION = "CORRELATION"
RULE = "RULE-BASED ASSESSMENT"
MODEL = "MODEL ESTIMATE"
UNKNOWN = "UNKNOWN"
PROVENANCE = (MEASURED, DERIVED, CORRELATION, RULE, MODEL, UNKNOWN)

# ── Data states: how complete a result is ───────────────────────────
OK = "OK"
NO_DATA = "NO DATA"
PARTIAL_DATA = "PARTIAL DATA"
NOT_MEASURED = "NOT MEASURED"
NOT_CONFIGURED = "NOT CONFIGURED"
INSUFFICIENT_HISTORY = "INSUFFICIENT HISTORY"
MODEL_NOT_VALIDATED = "MODEL NOT VALIDATED"
DATA_STATES = (OK, NO_DATA, PARTIAL_DATA, NOT_MEASURED, NOT_CONFIGURED,
               INSUFFICIENT_HISTORY, MODEL_NOT_VALIDATED)

# ── Refusals: the tool layer would not, or could not, answer ────────
NOT_PERMITTED = "NOT PERMITTED"
NOT_LICENSED = "NOT LICENSED"
NOT_FOUND = "NOT FOUND"
INVALID_ARGUMENTS = "INVALID ARGUMENTS"
FAILED = "FAILED"
REFUSALS = (NOT_PERMITTED, NOT_LICENSED, NOT_FOUND, INVALID_ARGUMENTS, FAILED)

# ── Likelihood: how a RULE rates something that has not happened yet ─
#
# Words, not probabilities. AMP has no calibrated forecast of a stoppage or a
# missed date, and a number would imply one; each risk states the rule and the
# threshold that earned its word (ai/risk_radar.py).
LIKELY = "LIKELY"
POSSIBLE = "POSSIBLE"
WATCH = "WATCH"
LIKELIHOOD = (LIKELY, POSSIBLE, WATCH)

# ── Root-cause labels (the Root-Cause Explorer) ─────────────────────
CAUSE_CONFIRMED = "CAUSE CONFIRMED"
LIKELY_CONTRIBUTOR = "LIKELY CONTRIBUTOR"
CORRELATED_EVENT = "CORRELATED EVENT"
INSUFFICIENT_EVIDENCE = "INSUFFICIENT EVIDENCE"
CAUSE_LABELS = (CAUSE_CONFIRMED, LIKELY_CONTRIBUTOR, CORRELATED_EVENT, INSUFFICIENT_EVIDENCE)


@dataclass(frozen=True)
class Fact:
    """One figure with its provenance.

    `key` is stable and machine-readable ("downtime.minutes"); `label` is what a
    person reads. `value` is a number, a string (a machine name, a reason) or
    None, and None is allowed ONLY with provenance UNKNOWN: a missing figure is
    stated as unknown, never shipped as a blank a reader fills in. A number must
    be finite, because NaN and infinity print as figures and mean nothing.
    """
    key: str
    label: str
    value: object
    provenance: str
    unit: str = ""
    source: str = ""
    window: str = ""
    detail: str = ""

    def __post_init__(self):
        if self.provenance not in PROVENANCE:
            raise ValueError(f"{self.key}: unknown provenance {self.provenance!r}")
        if self.value is None and self.provenance != UNKNOWN:
            raise ValueError(f"{self.key}: a missing value must be labelled {UNKNOWN}")
        if isinstance(self.value, bool):
            raise ValueError(f"{self.key}: a fact is a figure or a name, not a flag")
        if isinstance(self.value, float) and not math.isfinite(self.value):
            raise ValueError(f"{self.key}: {self.value!r} is not a figure")
        if not isinstance(self.value, (int, float, str, type(None))):
            raise ValueError(f"{self.key}: unsupported value type {type(self.value).__name__}")

    def to_dict(self, fact_id: str) -> dict:
        return {"id": fact_id, "key": self.key, "label": self.label, "value": self.value,
                "unit": self.unit, "provenance": self.provenance, "source": self.source,
                "window": self.window, "detail": self.detail}


@dataclass
class ToolResult:
    """What one AMP tool returned: a state, a sentence AMP wrote from the facts,
    the facts themselves and the screen that owns the detail."""
    tool: str
    state: str
    summary: str
    facts: list = field(default_factory=list)
    view: str = None
    notes: list = field(default_factory=list)
    elapsed_ms: int = None

    def __post_init__(self):
        if self.state not in DATA_STATES and self.state not in REFUSALS:
            raise ValueError(f"{self.tool}: unknown state {self.state!r}")

    @property
    def refused(self) -> bool:
        return self.state in REFUSALS

    def to_dict(self, first_id: int = 1) -> dict:
        return {"tool": self.tool, "state": self.state, "summary": self.summary,
                "view": self.view, "notes": list(self.notes), "elapsed_ms": self.elapsed_ms,
                "facts": [f.to_dict(f"F{first_id + i}") for i, f in enumerate(self.facts)]}


def refusal(tool: str, state: str, why: str) -> ToolResult:
    """A refusal as a result: no facts, one plain sentence saying why."""
    if state not in REFUSALS:
        raise ValueError(f"{state!r} is not a refusal")
    return ToolResult(tool=str(tool)[:80], state=state, summary=why)

File: frontend/lib/evidence.ts
// The Copilot's evidence vocabulary (ADR-0022), mirrored from backend/ai/evidence.py.
//
// Every figure the Copilot shows arrives as a Fact that says how AMP knows it. The
// strings below are the backend's, character for character, and
// backend/test_copilot_evidence_vocabulary.py fails the build if the two lists drift:
// a provenance the UI does not know would render as a blank chip, and a state it
// does not know would hide "NOT CONFIGURED" behind a normal-looking answer.
import { CURRENCY, money } from "./money";

export const PROVENANCE = [
  "MEASURED FACT",
  "DERIVED METRIC",
  "CORRELATION",
  "RULE-BASED ASSESSMENT",
  "MODEL ESTIMATE",
  "UNKNOWN",
] as const;
export type Provenance = (typeof PROVENANCE)[number];

export const DATA_STATES = [
  "OK",
  "NO DATA",
  "PARTIAL DATA",
  "NOT MEASURED",
  "NOT CONFIGURED",
  "INSUFFICIENT HISTORY",
  "MODEL NOT VALIDATED",
] as const;

export const REFUSALS = ["NOT PERMITTED", "NOT LICENSED", "NOT FOUND", "INVALID ARGUMENTS", "FAILED"] as const;

// Words, not probabilities: AMP has no calibrated forecast, and a number would
// imply one. Each risk states the rule that earned its word (ADR-0026).
export const LIKELIHOOD = ["LIKELY", "POSSIBLE", "WATCH"] as const;

export const CAUSE_LABELS = [
  "CAUSE CONFIRMED",
  "LIKELY CONTRIBUTOR",
  "CORRELATED EVENT",
  "INSUFFICIENT EVIDENCE",
] as const;

// What a Copilot answer may offer to PROPOSE (ADR-0039). A closed set, because
// each kind names a path AMP can actually carry out — a draft AMP cannot
// execute is a promise it cannot keep. Nothing here is a write: raising the
// draft is a separate authenticated call, and approving it is another.
export const PROPOSABLE_KINDS = ["maintenance_task"] as const;
export type ProposableKind = (typeof PROPOSABLE_KINDS)[number];

export type Fact = {
  id: string;
  key: string;
  label: string;
  value: number | string | null;
  unit: string;
  provenance: string;
  source: string;
  window: string;
  detail: string;
};

export type ToolRun = {
  tool: string;
  state: string;
  summary: string;
  notes?: string[];
  elapsed_ms?: number | null;
};

export type Grounding = { passed: boolean; numbers_checked: number; reasons: string[] } | null;

/**
 * An action a Copilot answer drafted, ready for a person to raise (ADR-0039).
 *
 * It is a description, not a record: nothing exists in AMP when this arrives.
 * Raising it sends only `kind` and `machine_id` — the server re-derives the
 * priority, the task type and the wording from the machine itself, so a field
 * edited here can never become what is written.
 */
export type Proposal = {
  kind: ProposableKind;
  machine_id: number;
  machine: string;
  priority: string;
  task_type: string;
  summary: string;
  reason: string;
  label: string;
};

/** What each provenance means, in one line a plant manager can read. */
export const PROVENANCE_MEANING: Record<Provenance, string> = {
  "MEASURED FACT": "Read directly from recorded data.",
  "DERIVED METRIC": "Calculated by a fixed formula from recorded data.",
  CORRELATION: "Two things moved together. Not proof that one caused the other.",
  "RULE-BASED ASSESSMENT": "A fixed rule or threshold. Not machine learning.",
  "MODEL ESTIMATE": "A trained model's estimate, with its validation status.",
  UNKNOWN: "AMP does not have this figure, and says so instead of guessing.",
};

export type Tone = "measured" | "derived" | "correlation" | "rule" | "model" | "unknown";

/** The chip colour family for a provenance; an unrecognised one reads as unknown. */
export function provenanceTone(p: string): Tone {
  switch (p) {
    case "MEASURED FACT":
      return "measured";
    case "DERIVED METRIC":
      return "derived";
    case "CORRELATION":
      return "correlation";
    case "RULE-BASED ASSESSMENT":
      return "rule";
    case "MODEL ESTIMATE":
      return "model";
    default:
      return "unknown";
  }
}

/**
 * A fact's value as it should read. Money only when the unit IS the currency the
 * backend priced it in; "unknown" for a missing value, never a zero or a blank.
 */
export function formatFactValue(f: Pick<Fact, "value" | "unit">): string {
  if (f.value === null || f.value === undefined) return "unknown";
  if (typeof f.value === "string") return f.value;
  if (!Number.isFinite(f.value)) return "unknown";
  if (f.unit === CURRENCY) return money(f.value);
  const n = f.value.toLocaleString();
  if (!f.unit) return n;
  if (f.unit === "%") return `${n}%`;
  if (f.unit.startsWith("/")) return `${n}${f.unit}`;
  return `${n} ${f.unit}`;
}

export function isRefusal(state: string | null | undefined): boolean {
  return !!state && (REFUSALS as readonly string[]).includes(state);
}

/**
 * The sentence that goes above an answer whose data is not complete. Null for OK,
 * so a normal answer carries no banner at all.
 */
export function stateNotice(state: string | null | undefined): string | null {
  switch (state) {
    case "NO DATA":
      return "No data recorded for this yet, so there is nothing to measure.";
    case "PARTIAL DATA":
      return "Partial data: part of the plant did not report, so this covers only what did.";
    case "NOT MEASURED":
      return "Not measured: AMP has no reading for this.";
    case "NOT CONFIGURED":
      return "Not configured: this needs a setting (a plan, a target or a unit value) before AMP can answer it.";
    case "INSUFFICIENT HISTORY":
      return "Not enough history yet to answer this fairly.";
    case "MODEL NOT VALIDATED":
      return "Model not validated on a real plant: treat this as an experimental estimate.";
    case "NOT PERMITTED":
      return "Your role can't see this.";
    case "NOT LICENSED":
      return "This is in a module pack your workspace doesn't include.";
    case "NOT FOUND":
      return "AMP couldn't find that in your workspace.";
    case "INVALID ARGUMENTS":
      return "AMP couldn't run that request as asked.";
    case "FAILED":
      return "A read failed, so nothing was answered from it.";
    default:
      return null;
  }
}
